# Carga y verificacion de datos

In [2]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np

In [3]:
zona_1 = pd.read_csv("geo_data_0.csv")
zona_2 = pd.read_csv("geo_data_1.csv")
zona_3 = pd.read_csv("geo_data_2.csv")

In [4]:
zona_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [5]:
zona_1.duplicated().sum()

np.int64(0)

In [6]:
zona_1.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

In [7]:
zona_1.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [8]:
zona_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [9]:
zona_2.duplicated().sum()

np.int64(0)

In [10]:
zona_2.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

In [11]:
zona_2.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [12]:
zona_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [13]:
zona_3.duplicated().sum()

np.int64(0)

In [14]:
zona_3.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

In [15]:
zona_3.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

Podemos ver que los datos de los tres datasets estan sin valores duplicados y sin datos ausentes por lo tanto podemos proceder a entrenar y probar el modelo que prediga el volumen de produccion de nuevos pozos en las tres regiones para identificar cual region es la mejor para realizar los 200 nuevos pozos

# Entrenamiento y prueba de modelo Regresion Lineal

Primero dividimos las columnas que nos brindan informacion para la creacion del modelo y las que no. Creamos una funcion para dividir los datos y entrenar el modelo, y asi obtener el RECM para observar que tan bien es el modelo para predecir y las predicciones de los modelos.

In [16]:
no_info_columns = ['id'] 
numeric_columns = ['f0','f1','f2','product']

def info_select (data,columns):
    zona_data_select = data.drop(columns,axis=1) #seleccionamos las columnas con datos relevantes para el modelo
    data_target = zona_data_select['product'] #tomamos los datos objetivo
    data_features = zona_data_select.drop('product', axis=1) #tomamos los datos caracteristicas 
    
    # Dividir los datos en conjuntos de entrenamiento y validación
    data_target_train, data_target_valid, data_features_train, data_features_valid = train_test_split(data_target, data_features, test_size=0.25, random_state=12345)

    # entrenamiento y validacion del modelo
    model = LinearRegression()
    model.fit(data_features_train,data_target_train)
    predictions_valid = model.predict(data_features_valid)

    #hallamos la raiz del error cuadratico medio
    result = mean_squared_error(data_target_valid,predictions_valid)**0.5 
    return result , predictions_valid

In [17]:
#llamamos la funcion info_select y obtenemos: la raiz del error cuadratico medio y las predicciones el volumen de reservas en pozos nuevos
result_1, predictions_valid_1 = info_select(zona_1,no_info_columns)
result_2, predictions_valid_2, = info_select(zona_2,no_info_columns)
result_3, predictions_valid_3, = info_select(zona_3,no_info_columns)

In [18]:
print("RECM del modelo de regresión lineal en el conjunto de validación de la zona_1 es:", result_1)
print("RECM del modelo de regresión lineal en el conjunto de validación de la zona_2 es:", result_2)
print("RECM del modelo de regresión lineal en el conjunto de validación de la zona_3 es", result_3)

RECM del modelo de regresión lineal en el conjunto de validación de la zona_1 es: 37.5794217150813
RECM del modelo de regresión lineal en el conjunto de validación de la zona_2 es: 0.8930992867756165
RECM del modelo de regresión lineal en el conjunto de validación de la zona_3 es 40.02970873393434


Zona_1: 37.5794217150813: Este valor es bastante alto, lo que indica que el modelo de regresión no está prediciendo con mucha precisión en esta zona. Puede haber varios factores que contribuyan a esto, como características geológicas específicas de la zona que no están siendo capturadas por el modelo, o posibles errores en los datos.

Zona_2: 0.8930992867756165: Este valor es muy bajo, lo que indica que el modelo de regresión está prediciendo con alta precisión en esta zona. Es probable que las características de los pozos en esta zona estén bien representadas por las variables del modelo.

Zona_3: 40.02970873393434: Este valor es incluso más alto que el de la Zona 1, lo que sugiere una baja precisión en las predicciones del modelo para esta zona. Al igual que en la Zona 1, pueden haber factores específicos de esta zona que el modelo no está capturando adecuadamente.

# Preparacion del cálculo de ganancias

Vamos a hallar la media de volumen predecido en cada region para ver si supera la media para no generar perdidas de 111.1 unidades (miles de barriles)

In [19]:
print("El volumen medio de la region 1 es:", predictions_valid_1.mean())
print("El volumen medio de la region 2 es:", predictions_valid_2.mean())
print("El volumen medio de la region 3 es:", predictions_valid_3.mean())

El volumen medio de la region 1 es: 92.59256778438035
El volumen medio de la region 2 es: 68.72854689544602
El volumen medio de la region 3 es: 94.96504596800489


Desafortunadamente con las predicciones podemos ver que las medias no superan la media para no generar perdidas en ninguna de las tres regiones en las que el modelo realizo predicciones de volumenes posibles para futuros pozos

Puede que con los 25000 datos predichos el volumen medio predecido no sea el que se pide de 111.1 unidades pero como se desea hacer una inversion de 100 millones por 200 pozos nuevos, vamos ha hacer un calculo de beneficios con los 200 pozos predichos que generan la mayor cantidad de unidades y ver si presentan beneficios despues de la inversion de 100 millones

# Cálculo de ganancias

In [20]:
# Creamos una funcion para calcular el posible beneficio neto de cada zona 
def profit(prediction):
    predictions_valid = pd.Series(prediction).nlargest(200) #pasamos los datos de prediccion de array de numpy a Series de pandas para facilitar el procesarlos
    net_profit = (predictions_valid * 4500).sum() - 100000000 #obtenemos el beneficio neto de cada zona al restarle la inversion de los pozos nuevos
    return net_profit

In [21]:
#LLamamos la funcion profit y hallamos los beneficios por zona de los 200 pozos con mayor volumen
net_profit_1 = profit(predictions_valid_1)
net_profit_2 = profit(predictions_valid_2)
net_profit_3 = profit(predictions_valid_3)

In [22]:
print('La ganancia potencia para la zona de pozos numero 1 es:',net_profit_1)
print('La ganancia potencia para la zona de pozos numero 2 es:',net_profit_2)
print('La ganancia potencia para la zona de pozos numero 3 es:',net_profit_3)

La ganancia potencia para la zona de pozos numero 1 es: 39960488.77465123
La ganancia potencia para la zona de pozos numero 2 es: 24857120.519735634
La ganancia potencia para la zona de pozos numero 3 es: 33217543.962432623


Como podemos ver las ganancias potenciales en cada region son las siguientes:

zona 1 = 39'960.488 USd

zona 2 = 24'857.120 USd

zona 3 = 33'217.543 USd 

Esto muestra a un claro ganador que es la zona 1 pero las predicciones dadas por el modelo me dejan con la duda si este seria el mejor sector para realizar los 200 nuevos pozos, lo mismo ocurre con la zona 3 que es el segundo puesto en potenciales ganancias. Desafortunadamente la zona con la menor potencial ganancia es la zona 2 la cual el modelo de regresion lineal mostro predecir con mayor eficiencia. 

Como conclusion la empresa no tendria perdida en ninguna de las zonas segun los valores dados por las predicciones, incluso en el peor de los escenarios que es en la zona 2.

# Cálculo Riesgos y Ganancias por región

In [23]:
def bootstrap_profit(prediction):
    state = np.random.RandomState(12345)
    values = []
    samples = 1000
    
    for i in range(samples):
        predictions = pd.Series(prediction) #pasamos los datos de prediccion de array de numpy a Series de pandas para facilitar el procesarlos
        subsample = predictions.sample(n=500,random_state=state,replace=True) #sacamos submuestras de 500 posos de las predicciones herchas por el modelo
        profit_200 = profit(subsample) #llamamos la funcion profit para obtener el beneficio de los 200 posos de mayor volumen de reserva predichos
        values.append(profit_200) #agregamos dichos beneficios a una lista para crear una distribucion de beneficios
    values = pd.Series(values) #convertimos la lista a una Serie para hallar el beneficio medio y otros datos de forma mas facil
    values_mean = values.mean() #hallamos el beneficio medio
    lower = values.quantile(0.025) #hallamos el minimo del intervalo de confianza o el beneficio minimo posible
    upper = values.quantile(0.975) #hallamos el maximo del intervalo de confianza o el beneficio mas alto posible
    loss_risk = (values < 0).mean() * 100 #halllamos el % de riego de perdida
    return values_mean, lower, upper, loss_risk


In [24]:
values_mean_1, lower_1, upper_1, loss_risk_1 = bootstrap_profit(predictions_valid_1)
values_mean_2, lower_2, upper_2, loss_risk_2 = bootstrap_profit(predictions_valid_2)
values_mean_3, lower_3, upper_3, loss_risk_3 = bootstrap_profit(predictions_valid_3)

In [25]:
print('El beneficio medio de la zona 1 es:',values_mean_1)
print('el valor minimo del intervalor de confianza de la zona 1 es:',lower_1)
print('el valor maximo del intervalor de confianza de la zona 1 es:',upper_1)
print('el riesgo de perdiad de la zona 1 es:', loss_risk_1)

El beneficio medio de la zona 1 es: 3580260.542083398
el valor minimo del intervalor de confianza de la zona 1 es: 1401373.3598106995
el valor maximo del intervalor de confianza de la zona 1 es: 5988772.713793823
el riesgo de perdiad de la zona 1 es: 0.1


In [26]:
print('El beneficio medio de la zona 2 es:',values_mean_2)
print('el valor minimo del intervalor de confianza de la zona 2 es:',lower_2)
print('el valor maximo del intervalor de confianza de la zona 2 es:',upper_2)
print('el riesgo de perdiad de la zona 2 es:', loss_risk_2)

El beneficio medio de la zona 2 es: 4538116.099461428
el valor minimo del intervalor de confianza de la zona 2 es: 328830.1962252781
el valor maximo del intervalor de confianza de la zona 2 es: 8545616.852130715
el riesgo de perdiad de la zona 2 es: 1.6


In [27]:
print('El beneficio medio de la zona 3 es:',values_mean_3)
print('el valor minimo del intervalor de confianza de la zona 3 es:',lower_3)
print('el valor maximo del intervalor de confianza de la zona 3 es:',upper_3)
print('el riesgo de perdiad de la zona 3 es:', loss_risk_3)

El beneficio medio de la zona 3 es: 2814661.9794109385
el valor minimo del intervalor de confianza de la zona 3 es: 920983.0189544179
el valor maximo del intervalor de confianza de la zona 3 es: 4785270.324733741
el riesgo de perdiad de la zona 3 es: 0.2


Como podemos ver en las tres zonas se presentan beneficios con los futuros 200 pozos, hay un riesgo de perdida mayor en la zona 2 con un 1,6% que podria asumirse aceptable ya que es la zona con mas beneficio pormedio de las 3 con 4'538.116 USD, esta zona la concidero la opcion mas viable no solo por los resultados de beneficios y riesgo aceptable, sino que tambien el modelo tubo muy bajo RECM lo que me dice que el modelo tiene muy buenas predicciones con respecto a los datos que nos otorgaron inicialmente. Las otras dos zonas a pesar de tener un riesgo muy bajo tienen menos beneficios y tambien tienen un RECM mayor lo que no me dejaria tomar una decision muy acertada del modelo. 